components/bus/message_bus.go

```diff
@@ -0,0 +1,119 @@
+package bus
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+type MessageSubscriberConstructor func(handlerName string) (message.Subscriber, error)
+
+type MessageBus struct {
+	router *message.Router
+
+	marshaler             MessageMarshaler
+	generateTopic         func(messageType string) string
+	subscriberConstructor MessageSubscriberConstructor
+
+	logger watermill.LoggerAdapter
+}
+
+func NewMessageBus(
+	router *message.Router,
+	marshaler MessageMarshaler,
+	generateTopic func(messageType string) string,
+	subscriberConstructor MessageSubscriberConstructor,
+	logger watermill.LoggerAdapter,
+) (*MessageBus, error) {
+	if router == nil {
+		return nil, errors.New("missing router")
+	}
+	if marshaler == nil {
+		return nil, errors.New("missing marshaler")
+	}
+	if generateTopic == nil {
+		return nil, errors.New("missing generateTopic")
+	}
+	if subscriberConstructor == nil {
+		return nil, errors.New("missing subscriberConstructor")
+	}
+	if logger == nil {
+		logger = watermill.NopLogger{}
+	}
+
+	return &MessageBus{
+		router,
+		marshaler,
+		generateTopic,
+		subscriberConstructor,
+		logger,
+	}, nil
+}
+
+func AddHandler[M any](
```

Comment:
shouldn't we maybe guard the new API with a build tag `//go:build go1.18` ?
And provide an alternative API for the previous versions?

Also, there are some concerns about the performance impact of generics, like https://planetscale.com/blog/generics-can-make-your-go-code-slower (I didn't really read the whole article carefully, don't have a strong opinion), so we could make it something like `//go:build go1.18 && generics` to make people enable generics consciously before we are comfortable with production use of generics-based code.

That being said, would it make sense to add a non-generic version? How would we attain that anyway, wouldn't we have to use `reflect` for example (which probably voids any concerns we may have over performance, doesn't it)?

Reply:
The Router and CQRS interfaces will still be there if people need it. The new API would be optional (but recommended).

Reply:
@maclav3 The article is good enough to explain some cases that you should careful to use generics in Go. And I recently reviewed the Go 1.20 changes that I belive Go 1.20 will fix some generics performance issues. E.g. https://github.com/golang/go/issues/54238. But in some cases, it's still be more significant performance degradation by design.